Trajectory specification for automotive-related TrajectoryAgents

jadecastro opened the pull request:
Relates #8529.

jadecastro commented:
+@stonier, for feature review.

---

Review status: 0 of 4 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/robotlocomotion/drake/8597#-:-LA0VaVs4VNmzlPSHlNQ:b-lbe63)*
<!-- Sent from Reviewable.io -->

stonier commented:
Review status: 0 of 4 files reviewed at latest revision, all discussions resolved.

---

*[a discussion](https://beta.reviewable.io/reviews/robotlocomotion/drake/8597#-LAmzNPD6jwHbIFOZMVy:-LAmzNPD6jwHbIFOZMVz:b-hvg3rr) (no related file):*
Working on the assumption here that the Trajectory class is a storage class only, with a simple interpolator. More interesting Trajectory generators will follow.

---

*[a discussion](https://beta.reviewable.io/reviews/robotlocomotion/drake/8597#-LAnBaRh64SHqHyNqJ4r:-LAnBaRh64SHqHyNqJ4s:blc0zuf) (no related file):*
Any reason the implementations are all in the header? It's not templated and it's not critical for speed - more useful to have it in the .cc files.

---

*[automotive/agent_trajectory.h, line 17 at r1](https://beta.reviewable.io/reviews/robotlocomotion/drake/8597#-LAmukY89w9TaSNe0gdC:-LAmukY89w9TaSNe0gdD:bxpbim1) ([raw file](https://github.com/robotlocomotion/drake/blob/d9519a2e02e90d05d36f3f71d7ab9712d17e5b69/automotive/agent_trajectory.h#L17)):*
> ```objc
> namespace automotive {
> 
> namespace internal {
> ```

An unnamed (anonymous) namespace will do here and be less verbose.

---

*[automotive/agent_trajectory.h, line 21 at r1](https://beta.reviewable.io/reviews/robotlocomotion/drake/8597#-LAn4vkg0QPTDgvjuEed:-LAn4vkg0QPTDgvjuEee:b-lbtt4m) ([raw file](https://github.com/robotlocomotion/drake/blob/d9519a2e02e90d05d36f3f71d7ab9712d17e5b69/automotive/agent_trajectory.h#L21)):*
> ```objc
> }  // namespace internal
> 
> using multibody::SpatialVelocity;
> ```

You don't want to impose these names stripped of their namespace prefix on anyone who includes the header file (fine if it was a .cc file). Drop these and make do with the full names  below.

---

*[automotive/agent_trajectory.h, line 31 at r1](https://beta.reviewable.io/reviews/robotlocomotion/drake/8597#-LAn8Uv-AJiMUwO3zmWB:-LAn8Uv03zTFQyOxH4iN:b-15zn6c) ([raw file](https://github.com/robotlocomotion/drake/blob/d9519a2e02e90d05d36f3f71d7ab9712d17e5b69/automotive/agent_trajectory.h#L31)):*
> ```objc
> /// coordinates, and translational and rotational velocities of frame A with
> /// respect to W.
> class AgentTrajectoryValues final {
> ```

Should be 'AgentTrajectoryValue'? It represents the snapshot at just one time instance and you do call it via `value()` in the AgentTrajectory class. It actually has no notion of trajectory, so could be renamed something else entirely.

---

*[automotive/agent_trajectory.h, line 58 at r1](https://beta.reviewable.io/reviews/robotlocomotion/drake/8597#-LAn8sms3xI8Y9Hs-rIW:-LAn8sms3xI8Y9Hs-rIX:b-skq09f) ([raw file](https://github.com/robotlocomotion/drake/blob/d9519a2e02e90d05d36f3f71d7ab9712d17e5b69/automotive/agent_trajectory.h#L58)):*
> ```objc
>   /// Accesses X_WA, an Isometry3 representation of pose of A, expressed in
>   /// world-frame coordinates W.
>   Eigen::Isometry3d isometry3() const {
> ```

Not using this anywhere here - will it be used frequently in followup PR's? Might be simple enough to just construct as needed elsewhere. Instead just having rotation() and translation() methods will make access faster rather than having to compose and decompose isometries all the time (as you do in the tests).

---

*[automotive/agent_trajectory.h, line 139 at r1](https://beta.reviewable.io/reviews/robotlocomotion/drake/8597#-LAn6S4c4lH9CEHSPMtq:-LAn6S4c4lH9CEHSPMtr:br9e1rz) ([raw file](https://github.com/robotlocomotion/drake/blob/d9519a2e02e90d05d36f3f71d7ab9712d17e5b69/automotive/agent_trajectory.h#L139)):*
> ```objc
>   // We friend AgentTrajectory to allow it to populate via the raw data vector
>   // constructor and raw data accessor.
>   friend class AgentTrajectory;
> ```

Is there a reason you need the 'friend' helpers in this class and not in the friend itself? If not, then decoupling the two classes and keeping this class minimal (~ struct) would be less entangled.

If it's a case of code re-use at some point, then it will have to be public or a non-member function anyway to avoid a growing list of friends.

---

*[automotive/agent_trajectory.h, line 170 at r1](https://beta.reviewable.io/reviews/robotlocomotion/drake/8597#-LAn410g0iuyNsLZTD94:-LAn410g0iuyNsLZTD95:bbmqa4q) ([raw file](https://github.com/robotlocomotion/drake/blob/d9519a2e02e90d05d36f3f71d7ab9712d17e5b69/automotive/agent_trajectory.h#L170)):*
> ```objc
>   /// @param knots a knot points containing a vector of AgentTrajectoryValues,
>   /// whose length must match that of @p times.
>   /// @throws std::logic_error if @p interp_type is not supported.
> ```

Worth mentioning here also the exceptions that PiecewisePolynomial will throw, e.g. if times are non-increasing.

---

*[automotive/agent_trajectory.h, line 171 at r1](https://beta.reviewable.io/reviews/robotlocomotion/drake/8597#-LAnAKnG6DrRMPJHutCU:-LAnAKnHEAyFvR5H-Dcd:b6ezanf) ([raw file](https://github.com/robotlocomotion/drake/blob/d9519a2e02e90d05d36f3f71d7ab9712d17e5b69/automotive/agent_trajectory.h#L171)):*
> ```objc
>   /// whose length must match that of @p times.
>   /// @throws std::logic_error if @p interp_type is not supported.
>   static AgentTrajectory Make(const InterpolationType& interp_type,
> ```

Move interp_type to the back of the arg list and give it a default? Users will often not be concerned about the interpolation type.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/robotlocomotion/drake/8597)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
Review status: 0 of 4 files reviewed at latest revision, 9 unresolved discussions.

---

*[a discussion](https://beta.reviewable.io/reviews/robotlocomotion/drake/8597#-LAnBaRh64SHqHyNqJ4r:-LAnx-1b1QnZhEuVcZkT:bt2vr1b) (no related file):*
<details><summary><i>Previously, stonier (Daniel Stonier) wrote…</i></summary><blockquote>

Any reason the implementations are all in the header? It's not templated and it's not critical for speed - more useful to have it in the .cc files.
</blockquote></details>

Fixed.

---

*[a discussion](https://beta.reviewable.io/reviews/robotlocomotion/drake/8597#-LAmzNPD6jwHbIFOZMVy:-LAnx8Ie2r63HZ1Jv3gD:b-78bcc8) (no related file):*
<details><summary><i>Previously, stonier (Daniel Stonier) wrote…</i></summary><blockquote>

Working on the assumption here that the Trajectory class is a storage class only, with a simple interpolator. More interesting Trajectory generators will follow.
</blockquote></details>

Yes. This just wraps PiecewisePolynomial so as to hide the underlying raw data structure, allowing callers to create trajectories and evaluate them using our own data structures.

---

*[automotive/agent_trajectory.h, line 17 at r1](https://beta.reviewable.io/reviews/robotlocomotion/drake/8597#-LAmukY89w9TaSNe0gdC:-LAnwq081I2cam9IlhPV:b-xkb2js) ([raw file](https://github.com/robotlocomotion/drake/blob/d9519a2e02e90d05d36f3f71d7ab9712d17e5b69/automotive/agent_trajectory.h#L17)):*
<details><summary><i>Previously, stonier (Daniel Stonier) wrote…</i></summary><blockquote>

An unnamed (anonymous) namespace will do here and be less verbose.
</blockquote></details>

Done. (Fixed now that .cc implementation is separate)

---

*[automotive/agent_trajectory.h, line 21 at r1](https://beta.reviewable.io/reviews/robotlocomotion/drake/8597#-LAn4vkg0QPTDgvjuEed:-LAnww31A6AboMqyw_W1:b-xkb2js) ([raw file](https://github.com/robotlocomotion/drake/blob/d9519a2e02e90d05d36f3f71d7ab9712d17e5b69/automotive/agent_trajectory.h#L21)):*
<details><summary><i>Previously, stonier (Daniel Stonier) wrote…</i></summary><blockquote>

You don't want to impose these names stripped of their namespace prefix on anyone who includes the header file (fine if it was a .cc file). Drop these and make do with the full names  below.
</blockquote></details>

Done. (Fixed now that .cc implementation is separate)

---

*[automotive/agent_trajectory.h, line 31 at r1](https://beta.reviewable.io/reviews/robotlocomotion/drake/8597#-LAn8Uv-AJiMUwO3zmWB:-LAnv6VK-2QibY80X7hc:b-rnuf0x) ([raw file](https://github.com/robotlocomotion/drake/blob/d9519a2e02e90d05d36f3f71d7ab9712d17e5b69/automotive/agent_trajectory.h#L31)):*
<details><summary><i>Previously, stonier (Daniel Stonier) wrote…</i></summary><blockquote>

Should be 'AgentTrajectoryValue'? It represents the snapshot at just one time instance and you do call it via `value()` in the AgentTrajectory class. It actually has no notion of trajectory, so could be renamed something else entirely.
</blockquote></details>

Done.  Renamed as `PoseVelocity`.

---

*[automotive/agent_trajectory.h, line 58 at r1](https://beta.reviewable.io/reviews/robotlocomotion/drake/8597#-LAn8sms3xI8Y9Hs-rIW:-LAnv2dR6d83wtjjpugR:bajjvxf) ([raw file](https://github.com/robotlocomotion/drake/blob/d9519a2e02e90d05d36f3f71d7ab9712d17e5b69/automotive/agent_trajectory.h#L58)):*
<details><summary><i>Previously, stonier (Daniel Stonier) wrote…</i></summary><blockquote>

Not using this anywhere here - will it be used frequently in followup PR's? Might be simple enough to just construct as needed elsewhere. Instead just having rotation() and translation() methods will make access faster rather than having to compose and decompose isometries all the time (as you do in the tests).
</blockquote></details>

Done.  I've made these into getters for `translation` and `rotation` (quaternions).

---

*[automotive/agent_trajectory.h, line 139 at r1](https://beta.reviewable.io/reviews/robotlocomotion/drake/8597#-LAn6S4c4lH9CEHSPMtq:-LAnuB4BDn6GiIAKvbfi:b-k2oja2) ([raw file](https://github.com/robotlocomotion/drake/blob/d9519a2e02e90d05d36f3f71d7ab9712d17e5b69/automotive/agent_trajectory.h#L139)):*
<details><summary><i>Previously, stonier (Daniel Stonier) wrote…</i></summary><blockquote>

Is there a reason you need the 'friend' helpers in this class and not in the friend itself? If not, then decoupling the two classes and keeping this class minimal (~ struct) would be less entangled.

If it's a case of code re-use at some point, then it will have to be public or a non-member function anyway to avoid a growing list of friends.
</blockquote></details>

Done.  I've removed the friending mechanism and now have the constructor and getter as public members with a warning in the docstring to prevent users from using them directly.  I'm keeping it a class because there's still some conversions (e.g. pose/speed conversions) we'd like to hide from the user. 

---

*[automotive/agent_trajectory.h, line 170 at r1](https://beta.reviewable.io/reviews/robotlocomotion/drake/8597#-LAn410g0iuyNsLZTD94:-LAnu9y511szeES_U6Bj:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d9519a2e02e90d05d36f3f71d7ab9712d17e5b69/automotive/agent_trajectory.h#L170)):*
<details><summary><i>Previously, stonier (Daniel Stonier) wrote…</i></summary><blockquote>

Worth mentioning here also the exceptions that PiecewisePolynomial will throw, e.g. if times are non-increasing.
</blockquote></details>

Done.

---

*[automotive/agent_trajectory.h, line 171 at r1](https://beta.reviewable.io/reviews/robotlocomotion/drake/8597#-LAnAKnG6DrRMPJHutCU:-LAnu90i6a6Q2bxxfx-a:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d9519a2e02e90d05d36f3f71d7ab9712d17e5b69/automotive/agent_trajectory.h#L171)):*
<details><summary><i>Previously, stonier (Daniel Stonier) wrote…</i></summary><blockquote>

Move interp_type to the back of the arg list and give it a default? Users will often not be concerned about the interpolation type.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/robotlocomotion/drake/8597)*
<!-- Sent from Reviewable.io -->

stonier commented:
Review status: 0 of 4 files reviewed at latest revision, 6 unresolved discussions.

---

*[automotive/agent_trajectory.h, line 21 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8597#-LAn4vkg0QPTDgvjuEed:-LArSzwmAmIprfke2jdI:b-wk4el0) ([raw file](https://github.com/robotlocomotion/drake/blob/d9519a2e02e90d05d36f3f71d7ab9712d17e5b69/automotive/agent_trajectory.h#L21)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

Done. (Fixed now that .cc implementation is separate)
</blockquote></details>

Ah, good ... forgot the detail that it was here in a header.

---

*[automotive/agent_trajectory.h, line 58 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8597#-LAn8sms3xI8Y9Hs-rIW:-LArV1gn0muWhTY9ylW0:b1usocy) ([raw file](https://github.com/robotlocomotion/drake/blob/d9519a2e02e90d05d36f3f71d7ab9712d17e5b69/automotive/agent_trajectory.h#L58)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

Done.  I've made these into getters for `translation` and `rotation` (quaternions).
</blockquote></details>

const references returns?

---

*[automotive/agent_trajectory.h, line 139 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8597#-LAn6S4c4lH9CEHSPMtq:-LArVGtq6PyodupG-k6S:b-lonkug) ([raw file](https://github.com/robotlocomotion/drake/blob/d9519a2e02e90d05d36f3f71d7ab9712d17e5b69/automotive/agent_trajectory.h#L139)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

Done.  I've removed the friending mechanism and now have the constructor and getter as public members with a warning in the docstring to prevent users from using them directly.  I'm keeping it a class because there's still some conversions (e.g. pose/speed conversions) we'd like to hide from the user. 
</blockquote></details>

Simply moving it to the Trajectory class would avoid having to have an awkward docstring warning.

---

*[automotive/agent_trajectory.h, line 170 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8597#-LAn410g0iuyNsLZTD94:-LArWCCi-PDLoWu0BstA:bdd2ts8) ([raw file](https://github.com/robotlocomotion/drake/blob/d9519a2e02e90d05d36f3f71d7ab9712d17e5b69/automotive/agent_trajectory.h#L170)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

It also throws if time is non-increasing right?

---

*[automotive/agent_trajectory.h, line 49 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8597#-LArW4k4EetOgLwI4Fn2:-LArW4k4EetOgLwI4Fn3:bha4ypm) ([raw file](https://github.com/robotlocomotion/drake/blob/156cd8e04780e15cdb7006c999fb6edbb901734c/automotive/agent_trajectory.h#L49)):*
> ```objc
>   ///
>   /// @param vector the vector of raw data values.
>   explicit PoseVelocity(const Eigen::VectorXd vector);
> ```

If you're going to have this, const ref input args  - http://drake.mit.edu/styleguide/cppguide.html#Use_of_const

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8597)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
Review status: 0 of 4 files reviewed at latest revision, 7 unresolved discussions.

---

*[automotive/agent_trajectory.h, line 58 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8597#-LAn8sms3xI8Y9Hs-rIW:-LAs5tID01rkbytp3G2D:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d9519a2e02e90d05d36f3f71d7ab9712d17e5b69/automotive/agent_trajectory.h#L58)):*
<details><summary><i>Previously, stonier (Daniel Stonier) wrote…</i></summary><blockquote>

const references returns?
</blockquote></details>

Done.

---

*[automotive/agent_trajectory.h, line 139 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8597#-LAn6S4c4lH9CEHSPMtq:-LAs5ujc3_Svs7mtDbMW:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d9519a2e02e90d05d36f3f71d7ab9712d17e5b69/automotive/agent_trajectory.h#L139)):*
<details><summary><i>Previously, stonier (Daniel Stonier) wrote…</i></summary><blockquote>

Simply moving it to the Trajectory class would avoid having to have an awkward docstring warning.
</blockquote></details>

Done.

---

*[automotive/agent_trajectory.h, line 170 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8597#-LAn410g0iuyNsLZTD94:-LAs635rEvw_ODcYHc7C:bsc4vvd) ([raw file](https://github.com/robotlocomotion/drake/blob/d9519a2e02e90d05d36f3f71d7ab9712d17e5b69/automotive/agent_trajectory.h#L170)):*
Yup. (and we have that documented below)

> ```
> /// `breaks` is not strictly increasing, and the inputs are otherwise
> ```

---

*[automotive/agent_trajectory.h, line 49 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8597#-LArW4k4EetOgLwI4Fn2:-LAs5tWRCsJIGBvF5Ov6:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/156cd8e04780e15cdb7006c999fb6edbb901734c/automotive/agent_trajectory.h#L49)):*
<details><summary><i>Previously, stonier (Daniel Stonier) wrote…</i></summary><blockquote>

If you're going to have this, const ref input args  - http://drake.mit.edu/styleguide/cppguide.html#Use_of_const
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8597)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
Review status: 0 of 4 files reviewed at latest revision, 6 unresolved discussions.

---

*[automotive/agent_trajectory.h, line 21 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8597#-LAn4vkg0QPTDgvjuEed:-LAs7-OQ59yhBF4MPW0M:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d9519a2e02e90d05d36f3f71d7ab9712d17e5b69/automotive/agent_trajectory.h#L21)):*
<details><summary><i>Previously, stonier (Daniel Stonier) wrote…</i></summary><blockquote>

Ah, good ... forgot the detail that it was here in a header.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8597)*
<!-- Sent from Reviewable.io -->

stonier commented:
*[automotive/agent_trajectory.h, line 115 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8597#-LAs9BuHD1NnSCUXoma4:-LAs9BuHD1NnSCUXoma5:b-fyp78q) ([raw file](https://github.com/robotlocomotion/drake/blob/7dfdd1b20476299c3afc11d9f7e1e1bbd7b633aa/automotive/agent_trajectory.h#L115)):*
> ```objc
>   /// @throws std::logic_error if @p interp_type is not supported.
>   /// @throws std::runtime_error if `times` and `knots` have different lengths,
>   /// `breaks` is not strictly increasing, and the inputs are otherwise
> ```

'times' not 'breaks'

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8597)*
<!-- Sent from Reviewable.io -->

stonier commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> pending resolution of the 'breaks' comment.

---







*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8597#-:-LAs9Mba8IwYxXTXQas6:b-d7sp5l)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
Review status: 0 of 4 files reviewed at latest revision, all discussions resolved.

---

*[automotive/agent_trajectory.h, line 115 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8597#-LAs9BuHD1NnSCUXoma4:-LAsTRfs9z3Ddx7owGMM:b7dmpfc) ([raw file](https://github.com/robotlocomotion/drake/blob/7dfdd1b20476299c3afc11d9f7e1e1bbd7b633aa/automotive/agent_trajectory.h#L115)):*
<details><summary><i>Previously, stonier (Daniel Stonier) wrote…</i></summary><blockquote>

'times' not 'breaks'
</blockquote></details>

Ah, thanks - good catch. Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8597)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
+@sammy-tri for platform review.

---

Review status: 0 of 4 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8597#-:-LAsTdfFDv2mvxrKyOLA:bpk00vl)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
First pass complete.  

I have a couple of questions though (and I can stop by to chat f2f if that's easier).

For the translational component...  The typical use case would be to store only the position, and call `derivative()` on the piecewise polynomial to get the velocity.  I think interpolating between the instantaneous velocities might give very different results depending on your use case (which I don't know).

For the rotational component, I'd expect to use something more like `PiecewiseQuaternionSlerp` which is intended to provide the right type of interpolation (it doesn't implement `derivative()`, but it does have it's own mechanism for providing angular velocity.

At any rate, it lot depends on what you're using this for.

---

Reviewed 2 of 4 files at r2, 1 of 2 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---

*[automotive/agent_trajectory.h, line 18 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8597#-LAsnhQtAgTC-wVYdM-F:-LAsnhQtAgTC-wVYdM-G:b-kzk15r) ([raw file](https://github.com/robotlocomotion/drake/blob/dee8ef53d1e6a5b35a4be44f35808960c6aa290e/automotive/agent_trajectory.h#L18)):*
> ```objc
> 
> /// Wraps the raw data contained in a trajectory expressed in terms of pose and
> /// velocitiy values.  Represents a translational and rotational transformation
> ```

typo "velocitiy"

---

*[automotive/agent_trajectory.h, line 46 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8597#-LAszGsz7FTKhsxuz8OJ:-LAszGsz7FTKhsxuz8OK:be0ytqm) ([raw file](https://github.com/robotlocomotion/drake/blob/dee8ef53d1e6a5b35a4be44f35808960c6aa290e/automotive/agent_trajectory.h#L46)):*
> ```objc
>   ///
>   /// Note that this constructor is for use within AgentTrajectory and callers
>   /// should NOT use it; instead use one of the above constructors.
> ```

The implementation for this appears to be missing?  (which is great if true since I didn't like having a public constructor we weren't supposed to use)

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8597#-:-LAt1-AG7i6ZXXXXT10y:b-x10pf3)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
Sure, @sammy-tri can you make it over here in the next hour?  We should catch @stonier too.

---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8597#-:-LAt96Wb6zuc31UBhdF_:bl8vlzh)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
Summary of f2f: I'll switch to storing pose only (zap velocity), and then store `rotation` in its own `PiecewiseQuaternionSlerp`. We can make use of `derivative()` to make a translational velocity, and `ComputeAngularVelocities()` for the rotational components.  I believe this makes things much nicer.

---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8597#-:-LAtJRTa81tJWetIpCNV:b-e0ov26)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
Changes made - thanks for the feedback @sammy-tri! 

---

Review status: 0 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[automotive/agent_trajectory.h, line 18 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8597#-LAsnhQtAgTC-wVYdM-F:-LAybuD7DsUEq5BHK9rK:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/dee8ef53d1e6a5b35a4be44f35808960c6aa290e/automotive/agent_trajectory.h#L18)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

typo "velocitiy"
</blockquote></details>

Done.

---

*[automotive/agent_trajectory.h, line 46 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8597#-LAszGsz7FTKhsxuz8OJ:-LAybvX_ASUC0RRknrO6:b2kkttj) ([raw file](https://github.com/robotlocomotion/drake/blob/dee8ef53d1e6a5b35a4be44f35808960c6aa290e/automotive/agent_trajectory.h#L46)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

The implementation for this appears to be missing?  (which is great if true since I didn't like having a public constructor we weren't supposed to use)
</blockquote></details>

Oops- removed the stray declaration.  Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8597#-:-LAyhBjGAAAcJsNfMGSO:b-poo2im)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>  (assuming CI is happy)

---

Reviewed 2 of 4 files at r5, 2 of 2 files at r6, 1 of 1 files at r7.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8597#-:-LAyiYsV9S1UlFvsq5-P:b1wkam9)*
<!-- Sent from Reviewable.io -->
